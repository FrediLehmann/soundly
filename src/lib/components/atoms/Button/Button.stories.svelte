<script>
  import { Meta, Story, Template } from '@storybook/addon-svelte-csf';
  import { userEvent, within } from '@storybook/testing-library';

  import Button from './Button.svelte';
  import { Settings } from '../../../Icons';
</script>

<Meta
  title="Atoms/Button"
  component={Button}
  argTypes={{
    href: {
      control: { type: 'text' },
      table: {
        type: {
          summary: 'If href is provided component renders as <a /> tag'
        }
      }
    },
    type: {
      control: { type: 'select' },
      options: ['button', 'submit', 'reset'],
      table: {
        defaultValue: {
          summary: 'button'
        },
        type: {
          summary: 'HTML button type.',
          detail:
            'Defines the HTML button type. This will change the behavior of the button. For forms use the submit type.'
        }
      }
    },
    style: {
      control: { type: 'select' },
      options: ['primary', 'secondary', 'ghost', 'link'],
      table: {
        defaultValue: {
          summary: 'secondary'
        },
        type: {
          summary: 'Decides how the button will be rendered.',
          detail:
            'There are different button style to render. Choose the correct styling based on your needs.'
        }
      }
    }
  }} />

<Template let:args>
  <div class="box">
    <Button {...args} type={args.type}>Button</Button>
  </div>
  <div class="box">
    <Button {...args} type={args.type} aria-label="settings">
      <Settings />
    </Button>
  </div>
  <div class="box">
    <Button {...args} type={args.type}>Button<Settings /></Button>
  </div>
  <div class="box">
    <Button {...args} type={args.type}><Settings />Button</Button>
  </div>
  <div class="box">
    <Button {...args} type={args.type} disabled>Button</Button>
  </div>
  <div class="box">
    <Button {...args} href="#" type={args.type}>Button as a tag</Button>
  </div>
</Template>

<Story
  name="Primary"
  args={{ style: 'primary' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const buttons = canvas.getAllByRole('button');

    buttons.forEach(async button => {
      userEvent.hover(button);
      userEvent.unhover(button);
      button.focus();
      button.blur();
    });
  }} />

<Story
  name="Secondary"
  args={{ style: 'secondary' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const buttons = canvas.getAllByRole('button');

    buttons.forEach(async button => {
      userEvent.hover(button);
      userEvent.unhover(button);
      button.focus();
      button.blur();
    });
  }} />

<Story
  name="Ghost"
  args={{ style: 'ghost' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const buttons = canvas.getAllByRole('button');

    buttons.forEach(async button => {
      userEvent.hover(button);
      userEvent.unhover(button);
      button.focus();
      button.blur();
    });
  }} />

<Story
  name="Link"
  args={{ style: 'link' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const buttons = canvas.getAllByRole('button');

    buttons.forEach(async button => {
      userEvent.hover(button);
      userEvent.unhover(button);
      button.focus();
      button.blur();
    });
  }} />

<style>
  .box {
    margin: 12px;
  }
</style>
