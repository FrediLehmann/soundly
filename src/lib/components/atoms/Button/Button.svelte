<script lang="ts">
  export let href: string = undefined;
  export let type: 'button' | 'submit' | 'reset' = 'button';
  export let style: 'primary' | 'secondary' | 'ghost' | 'link' | 'danger' =
    'secondary';
</script>

{#if href}
  <a {href} {type} data-type={style} {...$$restProps}>
    <slot />
  </a>
{:else}
  <button on:click {type} data-type={style} {...$$restProps}>
    <slot />
  </button>
{/if}

<style>
  a,
  button {
    text-decoration: none;
    box-sizing: border-box;

    display: inline-flex;
    justify-content: center;
    gap: 0.5rem;

    cursor: pointer;

    padding-block: 0.5rem;
    padding-inline: 0.75rem;

    border: none;
    border-radius: 0.125rem;

    font-weight: 600;

    align-items: center;
  }

  a:disabled,
  button:disabled {
    cursor: default;
    box-shadow: none;
    opacity: 0.7;
  }

  a[data-type='primary'],
  button[data-type='primary'] {
    color: hsla(var(--neutral-900));
    background-color: hsla(var(--primary));
  }

  a[data-type='primary']:hover:not(:disabled),
  button[data-type='primary']:hover:not(:disabled) {
    box-shadow: 0 0 0 2px hsla(var(--primary-400));
  }

  a[data-type='primary']:focus:not(:disabled),
  button[data-type='primary']:focus:not(:disabled),
  a[data-type='primary']:active:not(:disabled),
  button[data-type='primary']:active:not(:disabled) {
    box-shadow: 0 0 0 1px hsla(var(--neutral-000), 50%),
      0 0 0 3px hsla(var(--primary-400)), 0 0 #0000;
  }

  a[data-type='secondary'],
  button[data-type='secondary'] {
    color: hsla(var(--neutral-900), 100%);
    background-color: hsla(var(--neutral-400));
  }

  a[data-type='secondary']:hover:not(:disabled),
  button[data-type='secondary']:hover:not(:disabled) {
    box-shadow: 0 0 0 2px hsla(var(--neutral-200));
  }

  a[data-type='secondary']:focus:not(:disabled),
  button[data-type='secondary']:focus:not(:disabled),
  a[data-type='secondary']:active:not(:disabled),
  button[data-type='secondary']:active:not(:disabled) {
    box-shadow: 0 0 0 1px hsla(var(--neutral-000)),
      0 0 0 3px hsla(var(--neutral-200)), 0 0 #0000;
  }

  a[data-type='ghost'],
  button[data-type='ghost'] {
    background-color: hsla(var(--transparent), 0);
    color: hsla(var(--primary-700));
    border: 1px solid hsla(var(--primary-700));
  }

  a[data-type='ghost']:hover:not(:disabled),
  button[data-type='ghost']:hover:not(:disabled) {
    background-color: hsla(var(--primary-700));
    color: hsla(var(--neutral-000));
  }

  a[data-type='ghost']:focus:not(:disabled),
  button[data-type='ghost']:focus:not(:disabled),
  a[data-type='ghost']:active:not(:disabled),
  button[data-type='ghost']:active:not(:disabled) {
    box-shadow: 0 0 0 1px hsla(var(--neutral-000)),
      0 0 0 3px hsla(var(--primary-600)), 0 0 #0000;
  }

  a[data-type='danger'],
  button[data-type='danger'] {
    background-color: hsla(var(--transparent), 0);
    color: hsla(var(--secondary-500));
    border: 1px solid hsla(var(--secondary-500));
  }

  a[data-type='danger']:hover:not(:disabled),
  button[data-type='danger']:hover:not(:disabled) {
    background-color: hsla(var(--secondary-500));
    color: hsla(var(--neutral-000));
  }

  a[data-type='danger']:focus:not(:disabled),
  button[data-type='danger']:focus:not(:disabled),
  a[data-type='danger']:active:not(:disabled),
  button[data-type='danger']:active:not(:disabled) {
    box-shadow: 0 0 0 1px hsla(var(--neutral-000)),
      0 0 0 3px hsla(var(--seconary-600)), 0 0 #0000;
  }

  a[data-type='link'],
  button[data-type='link'] {
    background-color: hsla(var(--transparent));
    color: hsla(var(--primary-700));
  }
  a[data-type='link']:hover:not(:disabled),
  button[data-type='link']:hover:not(:disabled) {
    text-decoration: underline;
  }

  a,
  button[data-type='link']:focus:not(:disabled),
  a,
  button[data-type='link']:active:not(:disabled) {
    background-color: hsla(var(--primary-700), 10%);
  }
</style>
