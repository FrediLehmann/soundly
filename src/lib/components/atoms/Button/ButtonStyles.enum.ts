export enum ButtonStyles {
  Primary = 'my-auto px-5 py-2 h-full bg-blue-600 text-white hover:bg-blue-800 disabled:bg-blue-800 disabled:cursor-default',
  Secondary = 'my-auto px-5 py-2 h-full bg-gray-300 hover:bg-gray-400 disabled:bg-gray-400 disabled:cursor-default',
  Ghost = 'py-1 px-2 rounded-sm border border-blue-500 text-blue-500 hover:bg-blue-500 hover:text-white disabled:text-blue-500 disabled:bg-transparent disabled:cursor-default',
  Link = 'hover:underline'
}
