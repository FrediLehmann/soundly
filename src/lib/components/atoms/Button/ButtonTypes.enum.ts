export enum ButtonTypes {
    Primary = 'my-auto px-5 py-2 h-full bg-blue-600 text-white',
    Secondary = 'my-auto px-5 py-2 h-full bg-gray-300',
    Ghost = ''
}
