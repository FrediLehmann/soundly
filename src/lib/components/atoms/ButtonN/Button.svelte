<script type="ts">
  export let type: 'button' | 'submit' | 'reset' = 'button';
  export let style: 'primary' | 'secondary' | 'ghost' | 'link' = 'secondary';
</script>

<button on:click {type} data-type={style} {...$$restProps}>
  <slot />
</button>

<style>
  :root {
    --primary-color: hsl(204, 81%, 55%);
    --primary-shadow: hsla(204, 81%, 55%, 50%);
    --white: hsl(0, 0%, 100%);
  }

  button {
    cursor: pointer;

    padding-block: 0.5rem;
    padding-inline: 1.25rem;

    border: none;
    border-radius: 0.125rem;

    font-weight: 600;
  }

  button:disabled {
    cursor: default;
  }

  button[data-type='primary'] {
    color: var(--white);
    background-color: var(--primary-color);
  }

  button[data-type='primary']:hover {
    box-shadow: 0 0 0 2px var(--primary-shadow);
  }

  button[data-type='primary']:focus,
  button[data-type='primary']:active {
    box-shadow: 0 0 0 1px var(--white), 0 0 0 3px var(--primary-shadow),
      0 0 #0000;
  }

  button[data-type='primary']:disabled {
    box-shadow: none;
    opacity: 0.7;
  }
</style>
