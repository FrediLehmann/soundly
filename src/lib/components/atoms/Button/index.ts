import Button from './Button.svelte'
import { ButtonTypes } from './ButtonTypes.enum'

export { Button, ButtonTypes }