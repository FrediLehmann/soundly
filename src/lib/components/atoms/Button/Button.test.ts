/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom';
import { render } from "@testing-library/svelte";
import Button from "./Button.svelte";
import { ButtonTypes } from "./ButtonTypes.enum"

describe("Button component", () => {
    test("redners a button", () => {
        const { container } = render(Button);
        const button = container.querySelector('button')
        expect(button.className).toEqual(ButtonTypes.Secondary);
    });
});