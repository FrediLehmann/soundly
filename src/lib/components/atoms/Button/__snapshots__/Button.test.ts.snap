// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> component renders correctly 1`] = `
Slot {
  "$$": Object {
    "after_update": Array [],
    "before_update": Array [],
    "bound": Object {},
    "callbacks": Object {},
    "context": Map {},
    "ctx": Array [
      [Function],
      Object {},
    ],
    "dirty": Array [
      -1,
    ],
    "fragment": Object {
      "c": [Function],
      "d": [Function],
      "i": [Function],
      "l": [Function],
      "m": [Function],
      "o": [Function],
      "p": [Function],
    },
    "not_equal": [Function],
    "on_destroy": Array [
      [Function],
    ],
    "on_disconnect": Array [],
    "on_mount": Array [],
    "props": Object {
      "Component": 0,
    },
    "root": <div>
      <button
        class="svelte-19wuz13"
        data-type="secondary"
        type="button"
      >
        Button
      </button>
      
      
    </div>,
    "skip_bound": false,
    "update": [Function],
  },
  "$$set": [Function],
  "$capture_state": [Function],
  "$inject_state": [Function],
}
`;

exports[`<Button /> component renders correctly as link 1`] = `
Slot {
  "$$": Object {
    "after_update": Array [],
    "before_update": Array [],
    "bound": Object {},
    "callbacks": Object {},
    "context": Map {},
    "ctx": Array [
      [Function],
      Object {
        "href": "#",
      },
    ],
    "dirty": Array [
      -1,
    ],
    "fragment": Object {
      "c": [Function],
      "d": [Function],
      "i": [Function],
      "l": [Function],
      "m": [Function],
      "o": [Function],
      "p": [Function],
    },
    "not_equal": [Function],
    "on_destroy": Array [
      [Function],
    ],
    "on_disconnect": Array [],
    "on_mount": Array [],
    "props": Object {
      "Component": 0,
    },
    "root": <div>
      <a
        class="svelte-19wuz13"
        data-type="secondary"
        href="#"
        type="button"
      >
        Button
      </a>
      
      
    </div>,
    "skip_bound": false,
    "update": [Function],
  },
  "$$set": [Function],
  "$capture_state": [Function],
  "$inject_state": [Function],
}
`;
